### PR TITLE
Full typescript for WebGL code

### PR DESCRIPTION
**Pull request by ianthomas23**
This PR fully types the WebGL/ReGL code which previously contained many `any`s.  There is no associated issue but was last discussed/requested in PR #11159.  I have also removed the debug mesh renderer in `line_gl.ts` as it is no longer needed and certainly doesn't need typing.

The new `interface`s in `types.ts` are very useful; less so are the repeated
```typescript
u_canvas_size: regl.prop<Props, 'canvas_size'>('canvas_size'),
u_pixel_ratio: regl.prop<Props, 'pixel_ratio'>('pixel_ratio'),
u_antialias: regl.prop<Props, 'antialias'>('antialias'),
```
in `regl_wrap.ts`, but they are not difficult to understand or change so we may as well tolerate them.

I have tried not to make unrelated changes, so we are left with some duplication of code.  This will all be improved in the long run as the `interface` hierarchy will become cleaner when I implement missing functionality (e.g. hatching on markers), and the WebGL glyph member variables can be reduced/removed when I move some of the `regl_wrap.ts` functionality to the glyphs so that arrays are implemented as ReGL `Buffer`s to reduce the CPU to GPU communication and give much better performance.

### CODE REVIEW

File: bokehjs/src/lib/models/glyphs/webgl/regl_wrap.ts [hunk 1]
```diff
@@ -368,41 +323,45 @@ function regl_marker(regl: any, marker_type: MarkerType): ReglRenderFunction {
...
-        divisor: 1,
+      a_center: (_: Context, props: Props) => {
+        return {
+          buffer: regl.buffer(props.center),
+          divisor: 1,
+        }
       },
-      a_size: (_: any, props: any) => {
-        return one_each_or_constant(props.size, 1, false, props.nmarkers)
+      a_size: (_: Context, props: Props) => {
```
Comment (mattpap) [inline bokehjs/src/lib/models/glyphs/webgl/regl_wrap.ts:332]: All of those type annotations are unnecessary given you've already provided those types as generics to `DrawConfig`. Also you may want to consider using `a_size(_, props) { ... }` function shorthand syntax in objects.

File: bokehjs/src/lib/models/glyphs/webgl/types.ts [hunk 2]
```diff
@@ -0,0 +1,150 @@
+import {AttributeConfig, Texture2D, Vec2, Vec4} from "regl"
+
+
+interface CommonProps {
```
Comment (mattpap) [inline bokehjs/src/lib/models/glyphs/webgl/types.ts:4]: For the sake of consistency with the rest of bokehjs, we prefer `type` over `interface` in general, and use `interface` only were truly necessary.

File: bokehjs/src/lib/models/glyphs/webgl/types.ts [hunk 3]
```diff
@@ -0,0 +1,150 @@
+import {AttributeConfig, Texture2D, Vec2, Vec4} from "regl"
+
+
```
Comment (mattpap) [inline bokehjs/src/lib/models/glyphs/webgl/types.ts:3]: We don't use multiple empty lines in bokehjs.

File: bokehjs/src/lib/models/glyphs/webgl/types.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  nmarkers: number
+  width: Float32Array
+  height: Float32Array
+  angle: number[] | Float32Array
+  show: Uint8Array
+}
+
+export interface RectHatchGlyphProps extends RectGlyphProps, HatchProps {}
+
+
```
Comment (mattpap) [inline bokehjs/src/lib/models/glyphs/webgl/types.ts:71]: Once again multiple empty lines. If you see need for clear separation of section of module's code, then a comment would be preferable explaining following section's purpose.

File: bokehjs/src/lib/models/glyphs/webgl/regl_wrap.ts [hunk 5]
```diff
@@ -151,43 +144,50 @@ function regl_solid_line(regl: any): ReglRenderFunction {
...
     uniforms: {
-      u_canvas_size: regl.prop('canvas_size'),
-      u_pixel_ratio: regl.prop('pixel_ratio'),
-      u_color: regl.prop('color'),
-      u_linewidth: regl.prop('linewidth'),
-      u_antialias: regl.prop('antialias'),
-      u_miter_limit: regl.prop('miter_limit'),
-      u_join_type: regl.prop('join_type'),
-      u_cap_type: regl.prop('cap_type'),
+      u_canvas_size: regl.prop<Props, 'canvas_size'>('canvas_size'),
```
Comment (mattpap) [inline bokehjs/src/lib/models/glyphs/webgl/regl_wrap.ts:177]: This overspecification is distracting, but given how regl's API is design, it's the best that can be done, at least without adding an abstraction layer on top of the base regl API.

File: bokehjs/src/lib/models/glyphs/webgl/regl_wrap.ts (hunk 1 above)
Comment (ianthomas23) [reply bokehjs/src/lib/models/glyphs/webgl/regl_wrap.ts:332]: Having started adding types it is difficult to stop doing it everywhere, even if not really needed!

I've implemented both of your suggestions and the code is definitely simpler and easier to follow.  I did wonder if I needed to keep the `Context` types as they are not part of the `DrawConfig` generics, but there are no compiler complaints without them.

File: bokehjs/src/lib/models/glyphs/webgl/types.ts (hunk 2 above)
Comment (ianthomas23) [reply bokehjs/src/lib/models/glyphs/webgl/types.ts:4]: OK, I've switched from
```typescript
interface A extends B, C {... blah blah ...}
```
to
```typescript
type A = B & C & {... blah blah ...}
```

File: bokehjs/src/lib/models/glyphs/webgl/types.ts (hunk 4 above)
Comment (ianthomas23) [reply bokehjs/src/lib/models/glyphs/webgl/types.ts:71]: I've switched from 2 empty lines to 1, and added some comments where I think it helps.

Comment (ianthomas23) [reply bokehjs/src/lib/models/glyphs/webgl/types.ts:71]: Actually I've been putting double empty lines in many files, so I've reduced them all to single lines.

**Comment by mattpap** [comment]
Integration test failures are unrelated.

**Comment by bryevdv** [comment]
Thanks @ianthomas23 !

**Pull request merged**